### PR TITLE
 feat: integration on pinecone hosted rerankers

### PR DESCRIPTION
**Pull request by jzhao62**
# Description

- support 3 rerank model hosted by [pinecone-inference-model](https://app.pinecone.io/organizations/-Nn577_974iRsvC6nVxg/projects/a4fe57a4-b1cc-4a99-bf1d-c35a595cae4a/models)

- pinecone managed bge
- pinecone managed cohere
- pinecone managed native rerank algo
- 
## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No
image

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+nodes = [
+    NodeWithScore(node=TextNode(id_=f"vec{i}", text=txt))
+    for i, txt in enumerate(txts)
+]
+
+query_bundle = QueryBundle(
+    query_str="The tech company Apple is known for its innovative products like the iPhone."
+)
+
+reranker = PineconeNativeRerank(top_n=4, model="pinecone-rerank-v0")
```
Comment (logan-markewich) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md:32]: probably should show the api key here, or at least make a note that its set in the env

File: llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+- use rerank models with the pinecone managed vector service to rerank the search results
```
Comment (logan-markewich) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md:1]: This readme could use a header

File: llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    for i, txt in enumerate(txts)
+]
+
+query_bundle = QueryBundle(
+    query_str="The tech company Apple is known for its innovative products like the iPhone."
+)
+
+reranker = PineconeNativeRerank(top_n=4, model="pinecone-rerank-v0")
+
+result = reranker._postprocess_nodes(nodes=nodes, query_bundle=query_bundle)
```
Comment (logan-markewich) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md:34]: Really its `reranker.postprocess_nodes(...)` 

I guess this is how you use it directly, but most people will probably also throw it into a query engine

File: llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md (hunk 2 above)
Comment (jzhao62) [reply llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md:1]: resolved

File: llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md (hunk 3 above)
Comment (jzhao62) [reply llama-index-integrations/postprocessor/llama-index-postprocessor-pinecone-native-rerank/README.md]: you are right, i have updasted the read me with a more intuitive example. once the package is available in pip i will write the example as .jupyter

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**